src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Legacy/LegacySyntaxNodeExtensions.cs

```diff
@@ -1,97 +1,122 @@
+        SyntaxKind.RazorCommentTransition,
+        SyntaxKind.RazorCommentStar,
+        SyntaxKind.RazorCommentLiteral,
     };
 
-    private static readonly SyntaxKind[] CodeSpanKinds = new SyntaxKind[]
+    private static readonly ISet<SyntaxKind> s_codeSpanKinds = new HashSet<SyntaxKind>
     {
-            SyntaxKind.CSharpStatementLiteral,
-            SyntaxKind.CSharpExpressionLiteral,
-            SyntaxKind.CSharpEphemeralTextLiteral,
+        SyntaxKind.CSharpStatementLiteral,
+        SyntaxKind.CSharpExpressionLiteral,
+        SyntaxKind.CSharpEphemeralTextLiteral,
     };
 
-    private static readonly SyntaxKind[] MarkupSpanKinds = new SyntaxKind[]
+    private static readonly ISet<SyntaxKind> s_markupSpanKinds = new HashSet<SyntaxKind>
     {
-            SyntaxKind.MarkupTextLiteral,
-            SyntaxKind.MarkupEphemeralTextLiteral,
+        SyntaxKind.MarkupTextLiteral,
+        SyntaxKind.MarkupEphemeralTextLiteral,
     };
 
-    private static readonly SyntaxKind[] NoneSpanKinds = new SyntaxKind[]
+    private static readonly ISet<SyntaxKind> s_noneSpanKinds = new HashSet<SyntaxKind>
     {
-            SyntaxKind.UnclassifiedTextLiteral,
+        SyntaxKind.UnclassifiedTextLiteral,
     };
 
-    public static SpanContext GetSpanContext(this SyntaxNode node)
+    private static readonly ISet<SyntaxKind> s_allSpanKinds = CreateAllSpanKindsSet();
+
+    private static ISet<SyntaxKind> CreateAllSpanKindsSet()
     {
-        var context = node.GetAnnotationValue(SyntaxConstants.SpanContextKind);
+        var set = new HashSet<SyntaxKind>();
+
+        set.UnionWith(s_transitionSpanKinds);
+        set.UnionWith(s_metaCodeSpanKinds);
+        set.UnionWith(s_commentSpanKinds);
+        set.UnionWith(s_codeSpanKinds);
+        set.UnionWith(s_markupSpanKinds);
+        set.UnionWith(s_noneSpanKinds);
 
-        return context is SpanContext ? (SpanContext)context : null;
+        return set;
     }
 
-    public static TNode WithSpanContext<TNode>(this TNode node, SpanContext spanContext) where TNode : SyntaxNode
+    public static SpanContext? GetSpanContext(this SyntaxNode node)
+        => node.GetAnnotationValue(SyntaxConstants.SpanContextKind) as SpanContext;
+
+    public static TNode WithSpanContext<TNode>(this TNode node, SpanContext? spanContext)
+        where TNode : SyntaxNode
     {
-        if (node == null)
+        if (node is null)
         {
             throw new ArgumentNullException(nameof(node));
         }
 
         var newAnnotation = new SyntaxAnnotation(SyntaxConstants.SpanContextKind, spanContext);
 
-        List<SyntaxAnnotation> newAnnotations = null;
+        List<SyntaxAnnotation>? newAnnotations = null;
+
         if (node.ContainsAnnotations)
         {
-            var existingNodeAnnotations = node.GetAnnotations();
-            for (int i = 0; i < existingNodeAnnotations.Length; i++)
+            foreach (var annotation in node.GetAnnotations())
```

Comment:
This is a really complicated and inefficient way to either replace one element in an array, or allocate an array 1 element bigger and add the span context to it :).

Of note, though, is that SpanContext is one of the things that I'm working on eliminating entirely, so I'm not overly fussed if you don't want to make any more changes here.

Reply:
> This is a really complicated and inefficient way to either replace one element in an array, or allocate an array 1 element bigger and add the span context to it :).

I _totally_ agree. I decided to just do the minimum of clean up here because I knew you were going to jettison `SpanContext`.